apache/iceberg
# API,Core: Support Conditional Commits

# Context

Adds support for committing changes to an iceberg table based on whether or not a condition is true ***at*** commit time. 
Not before the commit. 
Not after the commit.
**_At_** commit time. 

This is useful in scenarios where users need a robust guard against potential concurrent commits. For example, some use cases require maintaining a monotonically increasing watermark in the snapshot properties. Our recently released iceberg-kafka-connect connector does this however it can only do this on a best-effort basis because Iceberg does not offer any API for expressing conditional commits. As a result, there is a risk of duplicate-file-appends there. This PR would enable closing that loophole. 

For more history/context/usecases, please see the discussion in https://github.com/apache/iceberg/issues/6514.

Incidentally, Delta (the competing table format) offers a similar feature albeit through a much more restricted API: https://github.com/delta-io/delta/blob/master/PROTOCOL.md#transaction-identifiers

# API design

We need to introduce a new API to allow users to declare the conditions under which a commit is allowed to proceed or not. There are two main options here: 
1. Add a new `void commitIf(List<Validation> validations)` method to the `PendingUpdate` interface. 
	1. See latest commit for implementation
	2. This offers a fluent-style API
2. Add a new `void validate(List<Validation> validations)` method to the `PendingUpdate` interface. 
	1. See first commit for implementation
	2. I feel this is simpler but the API is not as fluent as option 1 which some reviewers raised concerns about


# Note

- Please don't be put off by the size of the PR; 90% of it is purely tests.
